src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/ReviewUnusedParameters.cs

```diff
@@ -70,62 +64,111 @@ public override void Initialize(AnalysisContext context)
+                    // Map from parameter to a bool indicating if the parameter is used or not.
+                    var parameterUsageMap = new ConcurrentDictionary<IParameterSymbol, bool>();
 
-                compilationStartContext.RegisterOperationBlockStartAction(startOperationBlockContext =>
-                {
-                    // We only care about methods.
-                    if (!(startOperationBlockContext.OwningSymbol is IMethodSymbol method))
+                    // Set of methods which are used as delegates.
+                    var methodsUsedAsDelegates = new ConcurrentDictionary<IMethodSymbol, bool>();
+
+                    // Add candidate parameters for methods.
+                    symbolStartContext.RegisterOperationBlockStartAction(startOperationBlockContext =>
                     {
-                        return;
-                    }
+                        if (startOperationBlockContext.OwningSymbol is IMethodSymbol method &&
+                            ShouldAnalyzeMethod(method, startOperationBlockContext, eventsArgSymbol, attributeSetForMethodsToIgnore))
+                        {
+                            AddParameters(method, parameterUsageMap);
+                        }
+                    });
+
+                    // Add candidate parameters for local functions.
+                    symbolStartContext.RegisterOperationAction(
+                        context => AddParameters(((ILocalFunctionOperation)context.Operation).Symbol, parameterUsageMap),
+                        OperationKind.LocalFunction);
+
+                    // Add methods used as delegates.
+                    symbolStartContext.RegisterOperationAction(
+                        context => methodsUsedAsDelegates.TryAdd(((IMethodReferenceOperation)context.Operation).Method.OriginalDefinition, true),
+                        OperationKind.MethodReference);
+
+                    // Mark parameters with a parameter reference as used.
+                    symbolStartContext.RegisterOperationAction(
+                        context => parameterUsageMap.AddOrUpdate(
+                            ((IParameterReferenceOperation)context.Operation).Parameter,
+                            addValue: true,
+                            updateValueFactory: ReturnTrue),
+                        OperationKind.ParameterReference);
+
+                    // Report unused parameters in SymbolEnd action.
+                    symbolStartContext.RegisterSymbolEndAction(
+                        context => ReportUnusedParameters(context, parameterUsageMap, methodsUsedAsDelegates));
+                }, SymbolKind.NamedType);
+            });
+        }
 
-                    AnalyzeMethod(method, startOperationBlockContext, unusedMethodParameters,
-                        eventsArgSymbol, methodsUsedAsDelegates, attributeSetForMethodsToIgnore);
+        private static bool ReturnTrue(IParameterSymbol param, bool value) => true;
```

Comment:
Is it worth reviewing the other implementations to always create a static method instead of a local lambda?

Reply:
@Evangelink - yes. If you are interested, can you create a separate PR for this change? Please make sure that any changes to security rules are targeting 2.9.x branch.

Reply:
I will try to have a look but I want to finish the wildcard option for the inheritance tree ticket.